apps/src/lib/node/ledger/shell/vote_extensions/bridge_pool_vext.rs

```diff
@@ -0,0 +1,602 @@
+        let sig = Signed::<Vec<u8>, SignableEthBytes>::new(
+            shell.mode.get_eth_bridge_keypair().expect("Test failed"),
+            to_sign,
+        )
+        .sig;
+        let vote_ext = bridge_pool_roots::Vext {
+            block_height: shell.storage.get_current_decision_height(),
+            validator_addr: address,
+            sig,
+        }
+        .sign(shell.mode.get_protocol_key().expect("Test failed"));
+        let valid = shell
+            .filter_invalid_bp_roots_vexts(vec![
+                vote_ext.clone(),
+                vote_ext.clone(),
+            ])
+            .map(|(_, vext)| vext)
+            .collect::<Vec<_>>();
+        assert_eq!(valid, vec![vote_ext]);
+    }
+
+    /// Test that Bridge pool roots signed by a non-validator are rejected
+    /// even if the vext is signed by a validator
+    #[test]
+    fn test_bp_roots_must_be_signed_by_validator() {
+        let (shell, _, _) = setup_at_height(3u64);
+        let signing_key = gen_keypair();
+        let address = shell
+            .mode
+            .get_validator_address()
+            .expect("Test failed")
+            .clone();
+        let to_sign = [[0; 32], Uint::from(0).to_bytes()].concat();
+        let sig =
+            Signed::<Vec<u8>, SignableEthBytes>::new(&signing_key, to_sign).sig;
+        let bp_root = bridge_pool_roots::Vext {
+            block_height: shell.storage.get_current_decision_height(),
+            validator_addr: address,
+            sig,
+        }
+        .sign(shell.mode.get_protocol_key().expect("Test failed"));
+        assert!(!shell.validate_bp_roots_vext(
+            bp_root,
+            shell.storage.get_current_decision_height(),
+        ))
+    }
+
+    /// Test that Bridge pool root vext and inner signature
+    /// are from the same validator.
+    #[test]
+    fn test_bp_root_sigs_from_same_validator() {
+        let (mut shell, _broadcaster, _) = setup_at_height(3u64);
+        let address = shell
+            .mode
+            .get_validator_address()
+            .expect("Test failed")
+            .clone();
+        add_validator(&mut shell);
+        let to_sign = [[0; 32], Uint::from(0).to_bytes()].concat();
+        let sig = Signed::<Vec<u8>, SignableEthBytes>::new(
+            shell.mode.get_eth_bridge_keypair().expect("Test failed"),
+            to_sign,
+        )
+        .sig;
+        let bp_root = bridge_pool_roots::Vext {
+            block_height: shell.storage.get_current_decision_height(),
+            validator_addr: address,
+            sig,
+        }
+        .sign(&bertha_keypair());
+        assert!(!shell.validate_bp_roots_vext(
+            bp_root,
+            shell.storage.get_current_decision_height(),
+        ))
+    }
+
+    /// Test that an [`bridge_pool_roots::Vext`] that labels its included
+    /// block height as greater than the latest block height is rejected.
+    #[test]
+    fn reject_incorrect_block_number() {
```

Comment:
maybe refactor this test into a fn that accepts a block height, and add an additional ABCI++ specific test that checks if the ext is rejected based on the fact the inner height is lower than `last_height`

Reply:
In ABCI++, the bp_roots vext does not contain a block height, since there is only one valid block height. Thus this issue cannot occur.

Reply:
ah fair enough

Reply:
I lied. The block heights are there. I will add the tests.